File: stonequest/src/com/barelyconscious/game/spawnable/entities/SewerRatEntity.java
/* *****************************************************************************
   * File Name:         EntitySewerRat.java
   * Author:            Matt Schwartz
   * Date Created:      10.20.2012
   * Redistribution:    You are free to use, reuse, and edit any of the text in
                        this file.  You are not allowed to take credit for code
                        that was not written fully by yourself, or to remove 
                        credit from code that was not written fully by yourself.  
                        Please email schwamat@gmail.com for issues or concerns.
   * File Description:  
   ************************************************************************** */

package com.barelyconscious.game.spawnable.entities;

import com.barelyconscious.game.spawnable.Entity;
import com.barelyconscious.game.Common;
import com.barelyconscious.game.Game;
import com.barelyconscious.game.Sound;
import com.barelyconscious.game.World;
import com.barelyconscious.game.graphics.LineElement;
import com.barelyconscious.game.graphics.tiles.Tile;
import com.barelyconscious.game.item.Item;
import com.barelyconscious.game.spawnable.Loot;
import java.util.Random;

public class SewerRatEntity extends Entity {
    private final World world;
    private float minimumDamage = 0.75f;
    private float maximumDamage = 1.99f;
    private float criticalStrikeChance = 0.05f;
    private float criticalDamageMultiplier = 25f;
    
//    private long hitFreq = 1000;
//    private long last = System.currentTimeMillis();
    
    public SewerRatEntity(World world, int level, int x, int y) {
        super("Sewer Rat", Tile.SEWER_RAT_TILE_ID);
        super.setLevel(level);
        super.setHealthPoints(10f * (level * 0.76f));
        minimumDamage *= 1 + (level * 1.05);
        maximumDamage *= 1 + (level * 1.55);
        this.world = world;
        super.setPosition(x, y);
        super.setLastKnownPosition(Integer.MIN_VALUE, Integer.MIN_VALUE);
    } // constructor
    
    @Override
    public void tick() {
        
//        if (System.currentTimeMillis() < last + 1000) {
//            return;
//        }
//        
//        last = System.currentTimeMillis();
        
        int playerX = world.getPlayerX();
        int playerY = world.getPlayerY();
        int xDir = (playerX - getXPos()) >> 31;
        int yDir = (playerY - getYPos()) >> 31;
        int r = getReach() * Common.TILE_SIZE;
        
        /* Don't move if entity is in reach of player */
        if ( ((Math.abs(playerX - getXPos()) <= r) && (playerY - getYPos()) == 0) 
                || ((Math.abs(playerY - getYPos()) <= r) && (playerX - getXPos()) == 0) ) {
            interact();
            return;
        } // if
        
        // Pathing
        if (playerX > getXPos()) {
            xDir = 1;
        } // if
        if (playerY > getYPos()) {
            yDir = 1;
        } // if
        
        xDir *= Common.TILE_SIZE;
        yDir *= Common.TILE_SIZE;
        
        if (world.canMove(getXPos() + xDir, getYPos() + yDir) && !((getXPos() + xDir) == playerX && (getYPos() + yDir) == playerY)) {
            setPosition(getXPos() + xDir, getYPos() + yDir);
//            if (isVisible()) {
//                setLastKnownPosition(getXPos(), getYPos());
//            } // if
        } // if
        
        if (isVisible()) {
            setLastKnownPosition(getXPos(), getYPos());
        } // if
    } // tick
    
    @Override
    public void remove() {
        Loot drop;
        int goldTileId = Tile.GOLD_LOOT_STACK_TILE_ID;
        Random ran = new Random(System.currentTimeMillis());
        int amount = ran.nextInt(1400) + 1;
        
        amount *= super.getLevel();
        
        // Print death message and play sound
        Game.textLog.writeFormattedString(getDisplayName() + " dies.", Common.FONT_NULL_RGB, 
                new LineElement(getDisplayName(), true, Common.FONT_ENTITY_LABEL_RGB));
        
        // Drop any loot
        if (amount == 1) {
            goldTileId = Tile.GOLD_LOOT_SINGLE_TILE_ID; 
        } // if
        
        drop = new Loot(new Item("gold", 0, amount, goldTileId), getXPos(), getYPos());
        drop.setVisible(true);
        drop.setRemovableOnWalkover(true);
        
        world.addLoot(drop);
        
        super.remove();
    } // remove

    @Override
    public void interact() {
        float hit = calculateHit();
        Game.player.changeHealthBy(-hit);
        Game.textLog.writeFormattedString(getDisplayName() + " hits you for " + (int)hit + " physical.", 
                Common.FONT_DAMAGE_TEXT_RGB, new LineElement(getDisplayName(), true, 
                Common.FONT_ENTITY_LABEL_RGB));
        Sound.CHICKEN_CLUCK.play();
    } // interact
    
    private float calculateHit() {
        // Calculate a number between min hit and max hit
        Random ran = new Random();
        float hit =  minimumDamage;
        hit += (ran.nextFloat()) * (maximumDamage - minimumDamage);
        
        // Critical hit? multiply hit by critical multiplier
        if ((ran.nextDouble() * 100) <= criticalStrikeChance) {
            hit *= 1 + (criticalDamageMultiplier/100);
        } // if
        
        return hit;
    } // calculateHit
} // SewerRatEntity

File: StoneQuest/src/com/barelyconscious/game/spawnable/entities/SewerRatEntity.kt
package com.barelyconscious.game.spawnable.entities

import com.barelyconscious.game.Common
import com.barelyconscious.game.Game
import com.barelyconscious.game.Game.world
import com.barelyconscious.game.graphics.LineElement
import com.barelyconscious.game.graphics.tiles.Tile
import com.barelyconscious.game.item.Item
import com.barelyconscious.game.spawnable.Entity
import com.barelyconscious.game.spawnable.Loot
import com.barelyconscious.services.SoundService
import com.barelyconscious.systems.audio.PlayableSound
import com.barelyconscious.systems.MessageSystem
import com.barelyconscious.systems.messaging.data.SoundMessageData
import com.barelyconscious.services.TextLogMessageData
import com.barelyconscious.services.TextLogWriterService
import java.time.Clock
import java.util.*
import kotlin.math.abs
import kotlin.math.ceil

class SewerRatEntity(
    level: Int,
    x: Int,
    y: Int,
    messageSystem: MessageSystem
) : Entity("Sewer Rat", Tile.SEWER_RAT_TILE_ID, messageSystem) {

    private val minimumDamage: Double = 0.75 * (1 + (level * 1.05))
    private val maximumDamage: Double = 1.99 * (1 + (level * 1.55))
    private val criticalStrikeChance: Double = 0.05
    private val criticalDamageMultiplier: Double = 25.0

    init {
        super.level = level
        super.maxHealth = 10.0 * (level * 0.76)

        super.setPosition(x, y)
        super.setLastKnownPosition(Integer.MIN_VALUE, Int.MIN_VALUE)
    }

    override fun tick() {
        val reach = 1 // todo this was never really used

        val playerX: Int = world.playerX
        val playerY: Int = world.playerY
        var xDir: Int = (playerX - xPos) shr 31
        var yDir: Int = (playerY - yPos) shr 31
        val r: Int = reach * Common.TILE_SIZE

        if ((abs(playerX - xPos) <= r && (playerY - yPos) == 0)
            || (abs(playerY - yPos) <= r && (playerX - xPos) == 0)) {
            interact()
            return
        }

        if (playerX > xPos) {
            xDir = 1
        }
        if (playerY > yPos) {
            yDir = 1
        }

        xDir *= Common.TILE_SIZE
        yDir *= Common.TILE_SIZE

        val shouldMove = world.canMove(xPos + xDir, yPos + yDir)
            && !((xPos + xDir) == playerX && (yPos + yDir) == playerY)
        if (shouldMove) {
            setPosition(xPos + xDir, yPos + yDir)
        }

        if (isVisible) {
            setLastKnownPosition(xPos, yPos)
        }
    }

    override fun remove() {
        val drop: Loot
        val rand = Random(Clock.systemUTC().millis())
        val amount = (rand.nextInt(1400) + 1) * level

        messageSystem.sendMessage(
            TextLogWriterService.LOG_WRITE_TEXT,
            TextLogMessageData("$displayName has been murdered brutally. Its family mourns.")
                .with(LineElement(displayName, true, Common.FONT_ENTITY_LABEL_RGB)),
            this)

        drop = Loot(Item.createGoldItem(amount), xPos, yPos, messageSystem)
        drop.isVisible = true
        drop.removableOnWalkover = true

        world.addLoot(drop)

        super.remove()
    }

    override fun interact() {
        val hit: Double = calculateHit()

        // non-entity specific stuff that should be dealt with another way but who's keepin score?
        Game.player.changeHealthBy(-hit)
        messageSystem.sendMessage(
            TextLogWriterService.LOG_WRITE_TEXT,
            TextLogMessageData("$displayName hits you for ${ceil(hit).toInt()} physical")
                .with(LineElement(displayName, true, Common.FONT_ENTITY_LABEL_RGB)),
            this
        )

        messageSystem.sendMessage(
            SoundService.PLAY_SOUND,
            SoundMessageData(PlayableSound.CHICKEN_CLUCK),
            this)
    }

    private fun calculateHit(): Double {
        val rand = Random()
        val hit: Double = minimumDamage + rand.nextFloat() * (maximumDamage - minimumDamage)

        if ((rand.nextDouble() * 100) <= criticalStrikeChance) {
            return hit * (1 + criticalDamageMultiplier / 100)
        }

        return hit
    }

}
